### PR TITLE
[3.3.5] Unify chat link parsing

### PR DESCRIPTION
**Pull request by Treeston**
Follow-up to #22363, unify the chat link parsing and verification system.
By unify, I mean bring the new one up to par and then toss the old one.

Link type ~todo~ list:

- [x] `achievement`
- [x] `area`
- [x] `areatrigger`
- [x] `creature`
- [x] `creature_entry`
- [x] `enchant`
- [x] `gameevent`
- [x] `gameobject`
- [x] `gameobject_entry`
- [x] `glyph`
- [x] `item`
- [x] `itemset`
- [x] `player`
- [x] `quest`
- [x] `skill`
- [x] `spell`
- [x] `talent`
- [x] `taxinode`
- [x] `tele`
- [x] `title`
- [x] `trade`

### CODE REVIEW

**Comment by Treeston** [comment]
```
<Shauren> if i wanted to use this link stuff in legion
<Shauren> how would i change the hyperlink tokenizer to support "compressed" links
<Shauren> blizzard "compresses" links by removing zeros
<Shauren> |Hitem:2589:0:0:0:0:0:0:0:0 becomes |Hitem:2589::::::::
```

@Shauren change the `base_tag` parsers in `Hyperlinks.h` to parse integral values as zero for `len` of zero

(the Tokenizer in `HyperlinkTags.cpp` uses the parsers from `base_tag`)

File: src/server/game/Chat/Hyperlinks.cpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "Hyperlinks.h"
+#include "advstd.h"
+#include "Common.h"
+#include "DbcStores.h"
```
Comment (Shauren) [inline src/server/game/Chat/Hyperlinks.cpp]: case sensitive - DBCStores.h

File: src/server/game/Chat/Hyperlinks.cpp [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+        return false;
+    }
+};
+
+template <>
+struct LinkTextValidator<LinkTags::item>
+{
+    static bool IsValid(ItemLinkData const& data, char const* pos, size_t len)
+    {
+        ItemLocale const* locale = sObjectMgr->GetItemLocale(data.item->ItemId);
```
Comment (Shauren) [inline src/server/game/Chat/Hyperlinks.cpp]: this will fail if item with this id has no locale data in separate locale table (applies to every link type)

**Comment by Shauren** [comment]
Missing strict color validation

File: src/server/game/Chat/Hyperlinks.cpp (hunk 2 above)
Comment (Treeston) [reply src/server/game/Chat/Hyperlinks.cpp]: 👍

**Comment by ccrs** [comment]
Amazing work as always, you do love to work on such twisted logics, dont you? 😄

File: src/server/game/Chat/Hyperlinks.h [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    TalentEntry const* talent;
+    uint8 rank;
+};
+
+struct TradeskillLinkData
+{
+    SpellInfo const* spell;
+    uint16 curValue;
+    uint16 maxValue;
+    uint32 unk1;
```
Comment (Shauren) [inline src/server/game/Chat/Hyperlinks.h]: this is uint64, guid

File: src/server/game/Chat/HyperlinkTags.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    return true;
+}
+
+bool Trinity::Hyperlinks::LinkTags::trade::StoreTo(TradeskillLinkData& val, char const* pos, size_t len)
+{
+    HyperlinkDataTokenizer t(pos, len);
+    uint32 spellId;
+    if (!t.TryConsumeTo(spellId))
+        return false;
+    return (val.spell = sSpellMgr->GetSpellInfo(spellId)) && t.TryConsumeTo(val.curValue) && t.TryConsumeTo(val.maxValue) &&
```
Comment (Shauren) [inline src/server/game/Chat/HyperlinkTags.cpp]: should also check that spell has effect[0] == 47 (not any other effect slot)

File: src/server/game/Chat/Hyperlinks.h [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    std::string unk2;
```
Comment (Shauren) [inline src/server/game/Chat/Hyperlinks.h]: and this is encoded known recipe list

File: src/server/game/Chat/Hyperlinks.cpp [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+        return c == CHAT_LINK_COLOR_ACHIEVEMENT;
+    }
+};
+
+template <>
+struct LinkValidator<LinkTags::enchant>
+{
+    static bool IsTextValid(SpellInfo const* info, char const* pos, size_t len)
+    {
+        SkillLineAbilityMapBounds bounds = sSpellMgr->GetSkillLineAbilityMapBounds(info->Id);
```
Comment (Shauren) [inline src/server/game/Chat/Hyperlinks.cpp]: enchant links without related skill are allowed (client has code to construct them)
in which case the text is formatted as simply spellname, without "skillname: " prefix

File: src/server/game/Chat/Hyperlinks.cpp [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+            if (equal_with_len(name.c_str(), pos, len))
+                return true;
+        }
+
+        return false;
+    }
+
+    static bool IsColorValid(QuestLinkData const&, HyperlinkColor)
+    {
+        return true; // this color will be overridden clientside anyway
```
Comment (Shauren) [inline src/server/game/Chat/Hyperlinks.cpp]: should at least validate if it falls in allowed color set

Comment (Treeston) [reply src/server/game/Chat/Hyperlinks.cpp]: the client completely ignores the color specified here and replaces it for the appropriate difficulty color code for the current character

File: src/server/game/Chat/Hyperlinks.cpp (hunk 6 above)
Comment (Treeston) [reply src/server/game/Chat/Hyperlinks.cpp]: are you sure those aren't `spell:` links?

File: src/server/game/Chat/Hyperlinks.h (hunk 3 above)
Comment (Treeston) [reply src/server/game/Chat/Hyperlinks.h]: full guid of owner?

Comment (Treeston) [reply src/server/game/Chat/Hyperlinks.h]: (yes9

**Review by Shauren — APPROVED** [review]

**Pull request merged**